basic_obstacle_avoidance/basic_obstacle_avoidance.ino

```diff
@@ -37,15 +38,24 @@ unsigned long timestampMillis = 0;   // will store time after starting to perfor
 
 void setup() {
   Serial.begin(9600);
+  pinMode(8, OUTPUT);
   odometer.attach(odometerPin, [](){
     odometer.update();
   });
 }
 
 void loop() {
   check();
+  led(); 
 }
 
+void led(){
+  timestampMillis = millis();
+  while (millis() - timestampMillis < blinkInterval){
+      digitalWrite(8, LOW);
+    }
+  }
+
 void check(){
```

Comment:
What does this function do exactly?

Reply:
Also, don't use pin 8, it is already in use by the motors.

Reply:
Hi, 

I set a const int BlinkInterval of 600. So the method millis() is use as a timeStamp which is then compared to another timeStamp minus the BlinkInterval, thereby creating this interval instead of using delays.

Then, I set the LED pin to LOW to turn it off to create the blink effect during the intervals.

I saw pin 8 in use with the motors in the SmartCar code already, but it seems the Arduino board is able to recognize pin 8 for motors and pin 8 for the IC chip seperatedly upon figuring around with the code previously so that's why it was left on pin 8.

Anyway, the blinking code is just setup for now. I intend to reassign it to be used when the car is reversing and ither functionalities such as LED constantly on when moving forward or constantly off when parked or not moving etc.

Cheers~
Adelric

Reply:
Hi,
I am afraid your logic is fundamentally flawed. If it works, it does so **coincidentally** because at some point the motors which are also using this pin are setting it to HIGH.
What you have written in lines 53-54 is an spectacular alternative of `delay(blinkInterval);` :smile: 
Let's go through it line by line and you will see what I mean:
```cpp
  timestampMillis = millis();
  while (millis() - timestampMillis < blinkInterval){
      digitalWrite(8, LOW);
    }
```
The first line sets `timestampMillis` to some value, once. The second line, will **block** and keep setting the pin to LOW until approximately `blinkInterval` have elapsed. What you have is no different than:
```cpp
digitalWrite(8, LOW);
delay(blinkInterval);
```

To make things worse, I see that you are using this pattern [elsewhere](https://github.com/DIT112-V19/group-08/blob/09d8cbaae622d2c7796b82481bbf9ae03022277d/basic_obstacle_avoidance/basic_obstacle_avoidance.ino#L81-L93) too. Please refer to the [BlinkWithoutDelay](https://www.arduino.cc/en/Tutorial/BlinkWithoutDelay) example to see how `millis()` should be used in order to avoid blocking. Right now you have just created a strange `delay` equivalent. :crying_cat_face: 

Please revise ASAP. :+1:

Reply:
Hi,

Ooo I see, I understand what you mean.

Thanks for the critique but valuable feedback haha 👀

I will take a look at how to make the appropriate changes and amend it. I made something strange, yay! 😂

Cheers~

Reply:
Haha, don't see it as critique but as a well intended feedback. :smile:
Have fun!

Reply:
Yea no worries, it is definitely helpful! The detailed explanation gives me a clear and better understanding of what the issue is exactly. Thanks! 😊